File: adapters/go/adapter.go
// Rosetta reference adapter: Go.
//
// Implements the stdin/stdout JSON contract documented in
// docs/ADAPTER_CONTRACT.md. Reads a single JSON request object from stdin,
// writes a single JSON response object to stdout, and exits 0 even when the
// request itself is invalid (invalid input is reported via the "error"
// field, not a crash/non-zero exit).
//
// Supported operation: "convert" - converts an identifier between
// snake_case, camelCase, PascalCase and kebab-case.
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"
	"unicode"
)

type request struct {
	Operation string                 `json:"operation"`
	Input     string                 `json:"input"`
	Options   map[string]interface{} `json:"options"`
}

type response struct {
	Output *string `json:"output"`
	Error  *string `json:"error"`
}

var validStyles = map[string]bool{
	"snake": true, "camel": true, "pascal": true, "kebab": true,
}

// tokenize breaks an identifier into lowercase word tokens, regardless of
// its original case style.
func tokenize(identifier string) []string {
	identifier = strings.TrimSpace(identifier)

	var parts []string
	switch {
	case strings.Contains(identifier, "_"):
		parts = strings.Split(identifier, "_")
	case strings.Contains(identifier, "-"):
		parts = strings.Split(identifier, "-")
	default:
		// camelCase / PascalCase: insert a space before every interior
		// capital letter, then split on whitespace.
		var b strings.Builder
		for i, r := range identifier {
			if i > 0 && unicode.IsUpper(r) {
				b.WriteRune(' ')
			}
			b.WriteRune(r)
		}
		parts = strings.Split(b.String(), " ")
	}

	tokens := make([]string, 0, len(parts))
	for _, p := range parts {
		if p != "" {
			tokens = append(tokens, strings.ToLower(p))
		}
	}
	return tokens
}

func capitalize(s string) string {
	if s == "" {
		return s
	}
	r := []rune(s)
	r[0] = unicode.ToUpper(r[0])
	return string(r)
}

func joinTokens(tokens []string, style string) (string, error) {
	switch style {
	case "snake":
		return strings.Join(tokens, "_"), nil
	case "kebab":
		return strings.Join(tokens, "-"), nil
	case "camel":
		var b strings.Builder
		for i, t := range tokens {
			if i == 0 {
				b.WriteString(t)
			} else {
				b.WriteString(capitalize(t))
			}
		}
		return b.String(), nil
	case "pascal":
		var b strings.Builder
		for _, t := range tokens {
			b.WriteString(capitalize(t))
		}
		return b.String(), nil
	default:
		return "", fmt.Errorf("unknown case style: %s", style)
	}
}

func errorResponse(msg string) response {
	return response{Output: nil, Error: &msg}
}

func handle(req request) response {
	if req.Operation != "convert" {
		op := req.Operation
		if op == "" {
			op = "null"
		}
		return errorResponse(fmt.Sprintf("unsupported operation: %s", op))
	}
	if req.Input == "" {
		return errorResponse("input must be a non-empty string")
	}

	to, _ := req.Options["to"].(string)
	from, _ := req.Options["from"].(string)

	if !validStyles[to] {
		label := to
		if label == "" {
			label = "null"
		}
		return errorResponse(fmt.Sprintf("unsupported target case: %s", label))
	}
	if from != "" && !validStyles[from] {
		return errorResponse(fmt.Sprintf("unsupported source case: %s", from))
	}

	tokens := tokenize(req.Input)
	if len(tokens) == 0 {
		return errorResponse("could not tokenize input")
	}

	output, err := joinTokens(tokens, to)
	if err != nil {
		return errorResponse(err.Error())
	}
	return response{Output: &output, Error: nil}
}

func main() {
	raw, err := io.ReadAll(os.Stdin)

	var resp response
	if err != nil {
		resp = errorResponse(fmt.Sprintf("failed to read stdin: %v", err))
	} else {
		var req request
		if err := json.Unmarshal(raw, &req); err != nil {
			resp = errorResponse(fmt.Sprintf("invalid JSON payload: %v", err))
		} else {
			resp = handle(req)
		}
	}

	out, marshalErr := json.Marshal(resp)
	if marshalErr != nil {
		// Should be unreachable given the response shape, but keep the
		// contract intact even in that case.
		fmt.Fprintln(os.Stdout, `{"output":null,"error":"failed to marshal response"}`)
		return
	}
	fmt.Fprintln(os.Stdout, string(out))
}

File: adapters/php/adapter.php
<?php
/**
 * Rosetta reference adapter: PHP.
 *
 * Implements the stdin/stdout JSON contract documented in
 * docs/ADAPTER_CONTRACT.md. Reads a single JSON request object from stdin,
 * writes a single JSON response object to stdout, and exits 0 even when the
 * request itself is invalid (invalid input is reported via the `error`
 * field, not a crash/non-zero exit).
 *
 * Supported operation: "convert" - converts an identifier between
 * snake_case, camelCase, PascalCase and kebab-case.
 */

declare(strict_types=1);

const VALID_STYLES = ['snake', 'camel', 'pascal', 'kebab'];

/**
 * Break an identifier into lowercase word tokens, regardless of its
 * original case style.
 *
 * @return string[]
 */
function tokenize(string $identifier): array
{
    $identifier = trim($identifier);

    if (strpos($identifier, '_') !== false) {
        $parts = explode('_', $identifier);
    } elseif (strpos($identifier, '-') !== false) {
        $parts = explode('-', $identifier);
    } else {
        // camelCase / PascalCase: insert a space before every interior
        // capital letter, then split on whitespace.
        $spaced = preg_replace('/(?<!^)([A-Z])/', ' $1', $identifier);
        $parts = explode(' ', $spaced);
    }

    $tokens = [];
    foreach ($parts as $part) {
        if ($part !== '') {
            $tokens[] = strtolower($part);
        }
    }

    return $tokens;
}

function joinTokens(array $tokens, string $style): string
{
    switch ($style) {
        case 'snake':
            return implode('_', $tokens);
        case 'kebab':
            return implode('-', $tokens);
        case 'camel':
            $result = '';
            foreach ($tokens as $i => $token) {
                $result .= $i === 0 ? $token : ucfirst($token);
            }
            return $result;
        case 'pascal':
            $result = '';
            foreach ($tokens as $token) {
                $result .= ucfirst($token);
            }
            return $result;
        default:
            throw new InvalidArgumentException("Unknown case style: {$style}");
    }
}

function handle(array $request): array
{
    $operation = $request['operation'] ?? null;
    if ($operation !== 'convert') {
        throw new InvalidArgumentException(
            'Unsupported operation: ' . ($operation === null ? 'null' : (string) $operation)
        );
    }

    $input = $request['input'] ?? null;
    if (!is_string($input) || $input === '') {
        throw new InvalidArgumentException('Input must be a non-empty string.');
    }

    $options = $request['options'] ?? [];
    $to = is_array($options) ? ($options['to'] ?? null) : null;
    $from = is_array($options) ? ($options['from'] ?? null) : null;

    if (!in_array($to, VALID_STYLES, true)) {
        throw new InvalidArgumentException(
            'Unsupported target case: ' . ($to === null ? 'null' : (string) $to)
        );
    }
    if ($from !== null && !in_array($from, VALID_STYLES, true)) {
        throw new InvalidArgumentException("Unsupported source case: {$from}");
    }

    $tokens = tokenize($input);
    if (count($tokens) === 0) {
        throw new InvalidArgumentException('Could not tokenize input.');
    }

    return ['output' => joinTokens($tokens, $to), 'error' => null];
}

$raw = stream_get_contents(STDIN);
$response = ['output' => null, 'error' => null];

try {
    $request = json_decode((string) $raw, true, 512, JSON_THROW_ON_ERROR);
    if (!is_array($request)) {
        throw new InvalidArgumentException('Request payload must be a JSON object.');
    }
    $response = handle($request);
} catch (Throwable $e) {
    $response = ['output' => null, 'error' => $e->getMessage()];
}

fwrite(STDOUT, json_encode($response) . "\n");
